### DISCUSSION TITLE
[Idea] Allow to resend order confirmation email

### DISCUSSION THREAD

**Discussion by Hlavtox** [PrestaShop/PrestaShop#32249] [Ideas]
**Is your feature request related to a problem? Please describe.**
Sometimes, these use cases happen:
- User messes up his email and calls us, that he didn't receive his order confirmation. We change his email and now what?
- We change something in a customer's order and he wants the updated order summary.

**Describe the solution you'd like**
Add a native option to resend order confirmation email, as you could do with order status emails. 

Made it in-house by copying the functionality from PaymentModule class, but could be very useful to other PS users.

**Specs:**
Add a button to resend the order confirmation in the action bar in the header of the order or in the customer panel.

**Comment by khouloudbelguith** [comment]
Hi @Hlavtox,

We have an option in the BO => Orders => Order details page to resend an email
![image](https://user-images.githubusercontent.com/16067358/61470141-6fe59f80-a978-11e9-8865-029d84f2c05a.png)

Thanks to check & feedback.

**Comment by kpodemski** [comment]
@khouloudbelguith 

it's not possible though to re-send `order_conf.html` template which is send by `classes/PaymentModule.php` every time order is placed

this is why i understand this feature request :)

**Comment by Hlavtox** [comment]
@khouloudbelguith Not possible, order_conf is not a part of any order state.

**Comment by khouloudbelguith** [comment]
@Hlavtox, @kpodemski, Thanks for these clarifications.
Ping @marionf, @colinegin what do you think?

Thanks!

**Comment by MatShir** [comment]
@Hlavtox, it is missing and I saw in your order view design that you put a button on the general info, quite interesting. The feature is needed indeed, I will see it can be integrated into the redesign #14745.

**Comment by Hlavtox** [comment]
@MatShir Saw the InVision order screen design preview https://projects.invisionapp.com/share/YKTGGFZAMCB#/screens/381874234

I think the order resend button would nicely fit to the top row of buttons - next to print etc.

_If you made a hook there and modules could add their action there, even better. For example, for a payment gateway module to add a button which resends customer a link to pay/verify payment - or perhaps it could be done by displayAdminOrder and extending the correct smarty block?_

**Comment by MatShir** [comment]
> _If you made a hook there and modules could add their action there, even better. For example, for a payment gateway module to add a button which resends customer a link to pay/verify payment - or perhaps it could be done by displayAdminOrder and extending the correct smarty block?_

That seems a good idea. The hooks mapping is currently in the study. 
Just to let you know, we want to ship the feature but with the amount of time that design work and new feature of the page take us might reschedule for the next version. For now, we do our best and we let you know for any updates.

**Comment by madsoliver** [comment]
Commenting on this:
The most cases my clients has asked me "how do i resend" order_conf to customer, it is when the customer has mis-spelled his/her email.
Therefore: 1. Adding option to change email-address from order view and then 2. resend order_conf.

That would be great :)

**Comment by xmovu** [comment]
Hey guys,

Are there any news about this or an ETA?

It's definitely something we are missing. As well as being able to change the payment method to the client. If it can send the page to process the payment and allow to choose the gateway, it'd be perfect.

**Comment by Hlavtox** [comment]
@xmovu @onklmaps @kpodemski If you need this feature, here is my implementation. One link in admin order view and a function added to admin order controller. Adjust if needed. :-)

Insert this anywhere into `/www/admin527wi60rb/themes/default/template/controllers/orders/helpers/view/view.tpl`

```
<a href="{$current_index}&amp;submitResendConf&amp;id_order={$order->id}&amp;token={$smarty.get.token|escape:'html':'UTF-8'}"><i class="icon-mail-reply"></i>&nbsp;&nbsp;Resend order confirmation email</a>
```

Insert this into controllers/admin/AdminOrdersController.php, before `/* Update shipping number and carrier */`
```

/* Resend conf email */
        if (Tools::isSubmit('submitResendConf') && isset($order)) {
            if ($this->access('edit')) {
				
				$carrier = new Carrier((int)$order->id_carrier);
				$product_list_txt = '';
			        $product_list_html = '';
				$products = $order->getProducts();
				
                $customized_datas = Product::getAllCustomizedDatas((int)$order->id_cart);
				Product::addCustomizationPrice($products, $customized_datas);
				foreach ($products as $key => $product)
				{
					$unit_price = Product::getTaxCalculationMethod($customer->id) == PS_TAX_EXC ? $product['product_price'] : $product['product_price_wt'];

                    $customization_text = '';
                    if (isset($customized_datas[$product['product_id']][$product['product_attribute_id']])) {
                        foreach ($customized_datas[$product['product_id']][$product['product_attribute_id']][$order->id_address_delivery] as $customization) {
                            if (isset($customization['datas'][Product::CUSTOMIZE_TEXTFIELD])) {
                                foreach ($customization['datas'][Product::CUSTOMIZE_TEXTFIELD] as $text) {
                                    $customization_text .= $text['name'].': '.$text['value'].'<br />';
                                }
                            }

                            if (isset($customization['datas'][Product::CUSTOMIZE_FILE])) {
                                $customization_text .= count($customization['datas'][Product::CUSTOMIZE_FILE]).' '.$this->trans('image(s)', array(), 'Modules.Mailalerts.Admin').'<br />';
                            }
        
                            $customization_text .= '---<br />';
                        }
                        if (method_exists('Tools', 'rtrimString')) {
                            $customization_text = Tools::rtrimString($customization_text, '---<br />');
                        } else {
                            $customization_text = preg_replace('/---<br \/>$/', '', $customization_text);
                        }
                    }


					$url = $this->context->link->getProductLink($product['product_id']);
					$product_list_html .=
						'<tr>
                            <td style="padding: 10px 0;" width="18%">
                                <font size="2" face="Arial, sans-serif" color="#414a56">
                                    <p style="margin: 0; padding: 0 5px; font-size: 14px">'.$product['product_reference'].'</p>
                                </font>
                            </td>
                            <td style="padding: 10px 0;" width="47%">
                                <font size="2" face="Arial, sans-serif" color="#414a56">
                                    <p class="name-product" style="margin: 0; padding: 0 5px; font-size: 14px; color: #414a56; font-weight: bold;">'.$product['product_name']
                                    
									.(isset($product['attributes_small']) ? ' '.$product['attributes_small'] : '')
									.(!empty($customization_text) ? '<br />'.$customization_text : '')
									
								 . '</p>
                                </font>
                            </td>                            
                            <td style="padding: 10px 0;" width="15%">
                                <font size="2" face="Arial, sans-serif" color="#414a56">
                                    <p style="margin: 0; padding: 0 5px; font-size: 14px">'.Tools::displayPrice($unit_price, $currency, false).'</p>
                                </font>
                            </td>
                            <td style="padding: 10px 0;" width="5%">
                                <font size="2" face="Arial, sans-serif" color="#414a56">
                                    <p style="margin: 0; padding: 0 5px; font-size: 14px">'.(int)$product['product_quantity'].'</p>
                                </font>
                            </td>
                            <td align="right" style="padding: 10px 0;" width="15%">
                                <font size="2" face="Arial, sans-serif" color="#414a56">
                                    <p style="margin: 0; padding: 0 5px; font-size: 14px">' 
                                        .Tools::displayPrice(($unit_price * $product['product_quantity']), $currency, false)
                                    .'</p>
                                </font>
                            </td>
						</tr>';
						
					$product_list_txt .= $product['product_reference'].' - '.$product['product_name'].' - '	.(isset($product['attributes_small']) ? ' '.$product['attributes_small'] : '')
									.(!empty($customization_text) ? ' - '.$customization_text : '')
									
								.' - '.Tools::displayPrice($unit_price, $currency, false).' - '.(int)$product['product_quantity'].' ks - '
								.Tools::displayPrice(($unit_price * $product['product_quantity']), $currency, false);	
				}
				
				
				// Zjištění slev
				
				
				$cart_rules_list_txt = '';
                $cart_rules_list_html = '';
				
				foreach ($order->getCartRules() as $discount)
				{

                    $cart_rules_list_html .='<tr class="conf_body">
                    <td colspan="3" align="left" style="padding: 10px 0; ">
                        <font size="2" face="Arial, sans-serif" color="#414a56">
                            <p class="name-product" style="margin: 0; padding: 0 5px; font-size: 14px; color: #414a56; font-weight: bold;">'.$discount['name'].'</p>
                        </font>
                    </td>
                    <td colspan="2" align="right" style="padding: 10px 0; ">
                        <font size="2" face="Arial, sans-serif" color="#414a56">
                            <p style="margin: 0; padding: 0 5px; font-size: 14px">'.Tools::displayPrice($discount['value'], $currency, false).'</p>
                        </font>
                    </td>
                </tr>';
					
					$cart_rules_list_txt .=	$discount['name'].' - ' . Tools::displayPrice($discount['value'], $currency, false);
				}
							

				$invoice = new Address((int)$order->id_address_invoice);
                $delivery = new Address((int)$order->id_address_delivery);
                $delivery_state = $delivery->id_state ? new State((int)$delivery->id_state) : false;
                $invoice_state = $invoice->id_state ? new State((int)$invoice->id_state) : false;

				$data = array(
                            // openservis
                            '{message}' => $order->getFirstMessage(),
                            '{firstname}' => $this->context->customer->firstname,
                            '{lastname}' => $this->context->customer->lastname,
                            '{email}' => $this->context->customer->email,
                            '{delivery_block_txt}' => $this->_getFormatedAddress($delivery, "\n"),
                            '{invoice_block_txt}' => $this->_getFormatedAddress($invoice, "\n"),
                            '{delivery_block_html}' => $this->_getFormatedAddress($delivery, '<br />', array(
                                'firstname' => '<span style="font-weight:bold;">%s</span>',
                                'lastname' => '<span style="font-weight:bold;">%s</span>'
                            )),
                            '{invoice_block_html}' => $this->_getFormatedAddress($invoice, '<br />', array(
                                'firstname' => '<span style="font-weight:bold;">%s</span>',
                                'lastname' => '<span style="font-weight:bold;">%s</span>'
                            )),
                            '{delivery_company}' => $delivery->company,
                            '{delivery_firstname}' => $delivery->firstname,
                            '{delivery_lastname}' => $delivery->lastname,
                            '{delivery_address1}' => $delivery->address1,
                            '{delivery_address2}' => $delivery->address2,
                            '{delivery_city}' => $delivery->city,
                            '{delivery_postal_code}' => $delivery->postcode,
                            '{delivery_country}' => $delivery->country,
                            '{delivery_state}' => $delivery->id_state ? $delivery_state->name : '',
                            '{delivery_phone}' => ($delivery->phone) ? $delivery->phone : $delivery->phone_mobile,
                            '{delivery_other}' => $delivery->other,
                            '{invoice_company}' => $invoice->company,
                            '{invoice_vat_number}' => $invoice->vat_number,
                            '{invoice_firstname}' => $invoice->firstname,
                            '{invoice_lastname}' => $invoice->lastname,
                            '{invoice_address2}' => $invoice->address2,
                            '{invoice_address1}' => $invoice->address1,
                            '{invoice_city}' => $invoice->city,
                            '{invoice_postal_code}' => $invoice->postcode,
                            '{invoice_country}' => $invoice->country,
                            '{invoice_state}' => $invoice->id_state ? $invoice_state->name : '',
                            '{invoice_phone}' => ($invoice->phone) ? $invoice->phone : $invoice->phone_mobile,
                            '{invoice_other}' => $invoice->other,
                            '{order_name}' => $order->getUniqReference(),
                            '{date}' => Tools::displayDate($order->date_upd, null , 1),
                            // openservis
                            '{carrier}' => $carrier->name,
                            '{payment}' => Tools::substr($order->payment, 0, 32),
                            '{products}' => $product_list_html,
                            '{products_txt}' => $product_list_txt,
                            '{discounts}' => $cart_rules_list_html,
                            '{discounts_txt}' => $cart_rules_list_txt,
                            '{total_paid}' => Tools::displayPrice($order->total_paid, $this->context->currency, false),
                            '{total_paid_tax_excl}' => Tools::displayPrice($order->total_paid_tax_excl, $this->context->currency, false),
                            '{total_products}' => Tools::displayPrice(Product::getTaxCalculationMethod() == PS_TAX_EXC ? $order->total_products : $order->total_products_wt, $this->context->currency, false),
                            '{total_discounts}' => Tools::displayPrice($order->total_discounts, $this->context->currency, false),
                            '{total_shipping}' => Tools::displayPrice($order->total_shipping, $this->context->currency, false),
                            '{total_wrapping}' => Tools::displayPrice($order->total_wrapping, $this->context->currency, false),
                            '{total_tax_paid}' => Tools::displayPrice($order->total_paid_tax_incl - $order->total_paid_tax_excl, $this->context->currency, false));
				
				

                       if (Validate::isEmail($this->context->customer->email)) {
                            Mail::Send(
                                (int)$order->id_lang,
                                'order_conf',
                                Mail::l('Order confirmation', (int)$order->id_lang),
                                $data,
                                $this->context->customer->email,
                                $this->context->customer->firstname . ' ' . $this->context->customer->lastname,
                                null,
                                null,
                                null,
                                null, _PS_MAIL_DIR_, false, (int)$order->id_shop
                            );
                        }
				
				
				Tools::redirectAdmin(self::$currentIndex.'&id_order='.$order->id.'&vieworder&conf=10&token='.$this->token);	
            } else {
                $this->errors[] = $this->trans('You do not have permission to edit this.', array(), 'Admin.Notifications.Error');
            }
        }
/* Resend conf email */

	protected function _getFormatedAddress(Address $the_address, $line_sep, $fields_style = array())
    {
        return AddressFormat::generateAddress($the_address, array('avoid' => array()), $line_sep, ' ', $fields_style);
    }

```

**Comment by MatShir** [comment]
@Hlavtox it seems a nice work, you could contribute to the project if you have time. We will be pleased to review it 😄

**Comment by Hlavtox** [comment]
@MatShir No showcase of programming skills bro, just copied the functionality from emailalerts. :-D :-D

It could be done properly by creating "Order::sendConfirmationEmail" function, so we don't have duplicated code. Ping @matks for example?

**Comment by cowherdboy** [comment]
@Hlavtox I am trying to implement this but it is failing with error code 500 on this line:
'{delivery_block_txt}' => $this->_getFormatedAddress($delivery, "\n"),

What is the correct syntax for "$this->_getFormatedAddress" for PS 1.7.6.4?

**Comment by Hlavtox** [comment]
@cowherdboy Ooops, add this under the original code, sorry. Let me know if it works ;-)

```
protected function _getFormatedAddress(Address $the_address, $line_sep, $fields_style = array()) {
  return AddressFormat::generateAddress($the_address, array('avoid' => array()), $line_sep, ' ', $fields_style);
}
```

**Comment by cowherdboy** [comment]
@Hlavtox That did the trick. Thank you very much. My e-mail server went down for a few hours and I was getting payment notifications from my payment gateway but not the orders. Was needing this badly to send to customers.

**Comment by Jajcjusz** [comment]
> @xmovu @onklmaps @kpodemski If you need this feature, here is my implementation. One link in admin order view and a function added to admin order controller. Adjust if needed. :-)
> 
> Insert this anywhere into `/www/admin527wi60rb/themes/default/template/controllers/orders/helpers/view/view.tpl`
> 
> ```
> <a href="{$current_index}&amp;submitResendConf&amp;id_order={$order->id}&amp;token={$smarty.get.token|escape:'html':'UTF-8'}"><i class="icon-mail-reply"></i>&nbsp;&nbsp;Resend order confirmation email</a>
> ```
> 
> Insert this into controllers/admin/AdminOrdersController.php, before `/* Update shipping number and carrier */`
> 
> ```
> 
> /* Resend conf email */
>         if (Tools::isSubmit('submitResendConf') && isset($order)) {
>             if ($this->access('edit')) {
> 				
> 				$carrier = new Carrier((int)$order->id_carrier);
> 				$product_list_txt = '';
> 			        $product_list_html = '';
> 				$products = $order->getProducts();
> 				
>                 $customized_datas = Product::getAllCustomizedDatas((int)$order->id_cart);
> 				Product::addCustomizationPrice($products, $customized_datas);
> 				foreach ($products as $key => $product)
> 				{
> 					$unit_price = Product::getTaxCalculationMethod($customer->id) == PS_TAX_EXC ? $product['product_price'] : $product['product_price_wt'];
> 
>                     $customization_text = '';
>                     if (isset($customized_datas[$product['product_id']][$product['product_attribute_id']])) {
>                         foreach ($customized_datas[$product['product_id']][$product['product_attribute_id']][$order->id_address_delivery] as $customization) {
>                             if (isset($customization['datas'][Product::CUSTOMIZE_TEXTFIELD])) {
>                                 foreach ($customization['datas'][Product::CUSTOMIZE_TEXTFIELD] as $text) {
>                                     $customization_text .= $text['name'].': '.$text['value'].'<br />';
>                                 }
>                             }
> 
>                             if (isset($customization['datas'][Product::CUSTOMIZE_FILE])) {
>                                 $customization_text .= count($customization['datas'][Product::CUSTOMIZE_FILE]).' '.$this->trans('image(s)', array(), 'Modules.Mailalerts.Admin').'<br />';
>                             }
>         
>                             $customization_text .= '---<br />';
>                         }
>                         if (method_exists('Tools', 'rtrimString')) {
>                             $customization_text = Tools::rtrimString($customization_text, '---<br />');
>                         } else {
>                             $customization_text = preg_replace('/---<br \/>$/', '', $customization_text);
>                         }
>                     }
> 
> 
> 					$url = $this->context->link->getProductLink($product['product_id']);
> 					$product_list_html .=
> 						'<tr>
>                             <td style="padding: 10px 0;" width="18%">
>                                 <font size="2" face="Arial, sans-serif" color="#414a56">
>                                     <p style="margin: 0; padding: 0 5px; font-size: 14px">'.$product['product_reference'].'</p>
>                                 </font>
>                             </td>
>                             <td style="padding: 10px 0;" width="47%">
>                                 <font size="2" face="Arial, sans-serif" color="#414a56">
>                                     <p class="name-product" style="margin: 0; padding: 0 5px; font-size: 14px; color: #414a56; font-weight: bold;">'.$product['product_name']
>                                     
> 									.(isset($product['attributes_small']) ? ' '.$product['attributes_small'] : '')
> 									.(!empty($customization_text) ? '<br />'.$customization_text : '')
> 									
> 								 . '</p>
>                                 </font>
>                             </td>                            
>                             <td style="padding: 10px 0;" width="15%">
>                                 <font size="2" face="Arial, sans-serif" color="#414a56">
>                                     <p style="margin: 0; padding: 0 5px; font-size: 14px">'.Tools::displayPrice($unit_price, $currency, false).'</p>
>                                 </font>
>                             </td>
>                             <td style="padding: 10px 0;" width="5%">
>                                 <font size="2" face="Arial, sans-serif" color="#414a56">
>                                     <p style="margin: 0; padding: 0 5px; font-size: 14px">'.(int)$product['product_quantity'].'</p>
>                                 </font>
>                             </td>
>                             <td align="right" style="padding: 10px 0;" width="15%">
>                                 <font size="2" face="Arial, sans-serif" color="#414a56">
>                                     <p style="margin: 0; padding: 0 5px; font-size: 14px">' 
>                                         .Tools::displayPrice(($unit_price * $product['product_quantity']), $currency, false)
>                                     .'</p>
>                                 </font>
>                             </td>
> 						</tr>';
> 						
> 					$product_list_txt .= $product['product_reference'].' - '.$product['product_name'].' - '	.(isset($product['attributes_small']) ? ' '.$product['attributes_small'] : '')
> 									.(!empty($customization_text) ? ' - '.$customization_text : '')
> 									
> 								.' - '.Tools::displayPrice($unit_price, $currency, false).' - '.(int)$product['product_quantity'].' ks - '
> 								.Tools::displayPrice(($unit_price * $product['product_quantity']), $currency, false);	
> 				}
> 				
> 				
> 				// Zjištění slev
> 				
> 				
> 				$cart_rules_list_txt = '';
>                 $cart_rules_list_html = '';
> 				
> 				foreach ($order->getCartRules() as $discount)
> 				{
> 
>                     $cart_rules_list_html .='<tr class="conf_body">
>                     <td colspan="3" align="left" style="padding: 10px 0; ">
>                         <font size="2" face="Arial, sans-serif" color="#414a56">
>                             <p class="name-product" style="margin: 0; padding: 0 5px; font-size: 14px; color: #414a56; font-weight: bold;">'.$discount['name'].'</p>
>                         </font>
>                     </td>
>                     <td colspan="2" align="right" style="padding: 10px 0; ">
>                         <font size="2" face="Arial, sans-serif" color="#414a56">
>                             <p style="margin: 0; padding: 0 5px; font-size: 14px">'.Tools::displayPrice($discount['value'], $currency, false).'</p>
>                         </font>
>                     </td>
>                 </tr>';
> 					
> 					$cart_rules_list_txt .=	$discount['name'].' - ' . Tools::displayPrice($discount['value'], $currency, false);
> 				}
> 							
> 
> 				$invoice = new Address((int)$order->id_address_invoice);
>                 $delivery = new Address((int)$order->id_address_delivery);
>                 $delivery_state = $delivery->id_state ? new State((int)$delivery->id_state) : false;
>                 $invoice_state = $invoice->id_state ? new State((int)$invoice->id_state) : false;
> 
> 				$data = array(
>                             // openservis
>                             '{message}' => $order->getFirstMessage(),
>                             '{firstname}' => $this->context->customer->firstname,
>                             '{lastname}' => $this->context->customer->lastname,
>                             '{email}' => $this->context->customer->email,
>                             '{delivery_block_txt}' => $this->_getFormatedAddress($delivery, "\n"),
>                             '{invoice_block_txt}' => $this->_getFormatedAddress($invoice, "\n"),
>                             '{delivery_block_html}' => $this->_getFormatedAddress($delivery, '<br />', array(
>                                 'firstname' => '<span style="font-weight:bold;">%s</span>',
>                                 'lastname' => '<span style="font-weight:bold;">%s</span>'
>                             )),
>                             '{invoice_block_html}' => $this->_getFormatedAddress($invoice, '<br />', array(
>                                 'firstname' => '<span style="font-weight:bold;">%s</span>',
>                                 'lastname' => '<span style="font-weight:bold;">%s</span>'
>                             )),
>                             '{delivery_company}' => $delivery->company,
>                             '{delivery_firstname}' => $delivery->firstname,
>                             '{delivery_lastname}' => $delivery->lastname,
>                             '{delivery_address1}' => $delivery->address1,
>                             '{delivery_address2}' => $delivery->address2,
>                             '{delivery_city}' => $delivery->city,
>                             '{delivery_postal_code}' => $delivery->postcode,
>                             '{delivery_country}' => $delivery->country,
>                             '{delivery_state}' => $delivery->id_state ? $delivery_state->name : '',
>                             '{delivery_phone}' => ($delivery->phone) ? $delivery->phone : $delivery->phone_mobile,
>                             '{delivery_other}' => $delivery->other,
>                             '{invoice_company}' => $invoice->company,
>                             '{invoice_vat_number}' => $invoice->vat_number,
>                             '{invoice_firstname}' => $invoice->firstname,
>                             '{invoice_lastname}' => $invoice->lastname,
>                             '{invoice_address2}' => $invoice->address2,
>                             '{invoice_address1}' => $invoice->address1,
>                             '{invoice_city}' => $invoice->city,
>                             '{invoice_postal_code}' => $invoice->postcode,
>                             '{invoice_country}' => $invoice->country,
>                             '{invoice_state}' => $invoice->id_state ? $invoice_state->name : '',
>                             '{invoice_phone}' => ($invoice->phone) ? $invoice->phone : $invoice->phone_mobile,
>                             '{invoice_other}' => $invoice->other,
>                             '{order_name}' => $order->getUniqReference(),
>                             '{date}' => Tools::displayDate($order->date_upd, null , 1),
>                             // openservis
>                             '{carrier}' => $carrier->name,
>                             '{payment}' => Tools::substr($order->payment, 0, 32),
>                             '{products}' => $product_list_html,
>                             '{products_txt}' => $product_list_txt,
>                             '{discounts}' => $cart_rules_list_html,
>                             '{discounts_txt}' => $cart_rules_list_txt,
>                             '{total_paid}' => Tools::displayPrice($order->total_paid, $this->context->currency, false),
>                             '{total_paid_tax_excl}' => Tools::displayPrice($order->total_paid_tax_excl, $this->context->currency, false),
>                             '{total_products}' => Tools::displayPrice(Product::getTaxCalculationMethod() == PS_TAX_EXC ? $order->total_products : $order->total_products_wt, $this->context->currency, false),
>                             '{total_discounts}' => Tools::displayPrice($order->total_discounts, $this->context->currency, false),
>                             '{total_shipping}' => Tools::displayPrice($order->total_shipping, $this->context->currency, false),
>                             '{total_wrapping}' => Tools::displayPrice($order->total_wrapping, $this->context->currency, false),
>                             '{total_tax_paid}' => Tools::displayPrice($order->total_paid_tax_incl - $order->total_paid_tax_excl, $this->context->currency, false));
> 				
> 				
> 
>                        if (Validate::isEmail($this->context->customer->email)) {
>                             Mail::Send(
>                                 (int)$order->id_lang,
>                                 'order_conf',
>                                 Mail::l('Order confirmation', (int)$order->id_lang),
>                                 $data,
>                                 $this->context->customer->email,
>                                 $this->context->customer->firstname . ' ' . $this->context->customer->lastname,
>                                 null,
>                                 null,
>                                 null,
>                                 null, _PS_MAIL_DIR_, false, (int)$order->id_shop
>                             );
>                         }
> 				
> 				
> 				Tools::redirectAdmin(self::$currentIndex.'&id_order='.$order->id.'&vieworder&conf=10&token='.$this->token);	
>             } else {
>                 $this->errors[] = $this->trans('You do not have permission to edit this.', array(), 'Admin.Notifications.Error');
>             }
>         }
> /* Resend conf email */
> 
> 	protected function _getFormatedAddress(Address $the_address, $line_sep, $fields_style = array())
>     {
>         return AddressFormat::generateAddress($the_address, array('avoid' => array()), $line_sep, ' ', $fields_style);
>     }
> ```

How to integrate this in 1.6.2.4?

**Comment by cowherdboy** [comment]
@Hlavtox 
In PS 1.7.8.5 the following files are no longer there:
/adminxxxxx/themes/default/template/controllers/orders/helpers/view/view.tpl
and
/controllers/admin/AdminOrdersController.php

Any idea which files I should edit to implement your fix?

**Comment by wzzly** [comment]
I would like this (basic) feature as well. Should just be a button in the order details page to resend the confirmation mail to the customer...

**Comment by blastanders** [comment]
I just uploaded my module for re-sending order confirmation emails to my [personal Git](https://github.com/blastanders/Prestashop-Resend-Order-Confirmation-Email). Hope this will help somebody.

Im new to Prestashop's Git. If this breaks any rules, im happy to delete the link to my personal Git, or this comment.

**Reply by misieq12** [reply]
Hello,

Your module works fine but i Can resend e-mail only when order status = payment accepted. Can you are able to change it? 

So that I can resend the message for all order statuses? 

Regards.

**Reply by blastanders** [reply]
Hi mate, the module can only send order confirmation emails. If you want it to be able to send order confirmation emails for orders with other statuses, you can go to path_to_modules/presta_resend_order_conf_email/presta_resend_order_conf_email.php at around line 54, find "if ($order->current_state == '2') {" and remove that line, then remove the corresponding closing bracket "}" near line 62.

let me know if you don't have code access to your shop, happy to drop a modified version here for ya.

**Comment by cowherdboy** [comment]
@blastanders
I am using ps 1.7.8.5. The module installed fine but I do not see it appearing in the individual orders. How to troubleshoot the issue?

**Reply by blastanders** [reply]
Hey, there have been multiple people asking me questions related to yours. I have modified the code so that it shows the button regardless of the order status. Download the git again and re-install it. It should solve the problem you have. If not, let me know.

**Comment by darkf3d3** [comment]
Any chance to have this feature without the need to install yet another module?

**Comment by blastanders** [comment]
not that i know of. adding module is the easiest way to make changes to
your shop.

**Comment by MatShir** [comment]
This feature could clearly be a native feature. If someone want to implement it. We will be happy to review it ;)

**Reply by Hlavtox** [reply]
I tried to do it, but it cant be done easily now, until symfony container is implemented in FO

**Reply by blastanders** [reply]
Hi @MatShir . Sorry if this is a dumb question. When you say implement it, did you mean as a module, or modify the core code of prestashop to include this feature?

**Reply by MatShir** [reply]
When I speak as a native feature, I'm referring to add the feature into the core through a Pull Request on GitHub, so every one can have the feature when they install or upgrade to the latest PrestaShop version. But actually, It could also a feature included into native module. Meaning that the feature will be available for user in a module on new PS install and current user just have to upgrade the module.

**Comment by sharkooon** [comment]
Its a feature everyone would use if a customer entered a wrong email address. And that is happening allot.
Also for testing purposes it could be helpfull.

**Comment by HelixMartin** [comment]
I agree that this features has been needed for a long time. It happens all the time, customers wants a copy of the order confirmations, for different reasons.

**Comment by MariuszKa75** [comment]
Hello,
In prestashop 1.7.8.11 it doesn't work:

Symfony\Component\Debug\Exception\ContextErrorException:
Notice: Undefined variable: currency

  at modules/presta_resend_order_conf_email/controllers/admin/AdminPrestaResendOrderConfEmail.php:93
  at AdminPrestaResendOrderConfEmailController->ajaxProcessresendOrderConfEmail()
     (classes/controller/AdminController.php:983)
  at AdminControllerCore->postProcess()
     (modules/presta_resend_order_conf_email/controllers/admin/AdminPrestaResendOrderConfEmail.php:17)
  at AdminPrestaResendOrderConfEmailController->postProcess()
     (classes/controller/Controller.php:295)
  at ControllerCore->run()
     (classes/Dispatcher.php:525)
  at DispatcherCore->dispatch()

Any solutions?

Best regards
Mariusz

**Comment by fire2** [comment]
> presta_resend_order_conf_email

You're using a module. That is not part of Prestashop core. Issue should be reported to the module developer.
